### PR TITLE
Feature / Add breadcrumb component

### PR DESCRIPTION
**Pull request by Zyruks**
## Changes Made 🎉

- [x] feat(breadcrumb): ✨[add breadcrumb component](https://github.com/serudda/side-ui/commit/c0af871c9f70cbc62afcf396135ac5eece6c0275)
- [x] feat(story): ✨[add breadcrumb story](https://github.com/serudda/side-ui/commit/6481679e031850264b5106ea1f9da8cbd74895da)
- [x] fix(tailwind): 🐛[darkMode is bug without adding darkMode in config file](https://github.com/serudda/side-ui/commit/4a7e59f08a51eb1fda3c4e983ffa9c2d9bd2ad85)

### Describe Changes

This PR introduces the `Breadcrumb` component, aiming to improve the user navigation experience by providing a clear path back to the starting or entry point of a website or application. The addition of a Storybook story for the `Breadcrumb` component allows developers and designers to interact with the component and understand its functionality and customization options. Furthermore, a minor UI enhancement was made by updating the hover color for better visibility and interaction. Lastly, a critical fix was applied to the Tailwind configuration to support dark mode properly, ensuring a consistent and user-friendly theme across the application.

## Visuals (Optional)

https://github.com/serudda/side-ui/assets/14036522/d3a5c3fb-fc72-4744-a12b-eb621a78be59

## Checklist ✅

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: src/components/Breadcrumbs/utils/helper.tsx [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+import { cn } from '@common';
+import { CollapseMode } from '@components';
+
+export enum ItemIdentifier {
+  before = 'before',
+  collapse = 'collapse',
+  after = 'after',
+}
+
+interface ProcessBreadcrumbItemProps {
```
Comment (serudda) [inline src/components/Breadcrumbs/utils/helper.tsx:11]: Por que si es un util, lo defines como un component?

File: tailwind.config.js [hunk 2]
```diff
@@ -1,5 +1,39 @@
 module.exports = {
   //presets: [require('./tailwind-presets/storybook-preset')],
+  darkMode: 'class',
   content: ['./src/**/*.{js,ts,jsx,tsx}'],
+  theme: {
+    extend: {
+      colors: {
+        primary: {
+          50: '#EFF8FF',
```
Comment (serudda) [inline tailwind.config.js:9]: Estos colores de donde salieron?

**Review by serudda — CHANGES_REQUESTED** [review]

File: src/components/Breadcrumbs/Breadcrumb.stories.tsx [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+  render: (args) => {
+    const spacing = Object.values(BreadcrumbSpacing);
+
+    return (
+      <div>
+        <div className="grid gap-4">
+          {spacing.map((spacing) => (
+            <div key={spacing}>
+              <p>Spacing: {spacing}</p>
+              <Breadcrumb {...args} spacing={spacing}>
```
[the remark this answers is not in the document]
Comment (Zyruks) [reply src/components/Breadcrumbs/Breadcrumb.stories.tsx:53]: Spacing es el gap entre elementos.

¿Debería tener otro nombre?

image

File: src/components/Breadcrumbs/Breadcrumb.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+        <BreadcrumbItem href="javascript:void(0)">Calendar</BreadcrumbItem>
+      </Breadcrumb>
+    );
+  },
+};
+
+export const MoreItemsAfterCollapse: Story = {
+  render: (args) => {
+    return (
+      <Breadcrumb {...args} isCollapse={true} itemsBeforeCollapse={1} itemsAfterCollapse={2}>
```
[the remark this answers is not in the document]
Comment (Zyruks) [reply src/components/Breadcrumbs/Breadcrumb.stories.tsx:148]: Done!

File: src/components/Breadcrumbs/Breadcrumb.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+        <BreadcrumbItem href="javascript:void(0)">Calendar</BreadcrumbItem>
+      </Breadcrumb>
+    );
+  },
+};
+
+export const MoreItemsBeforeCollapse: Story = {
+  render: (args) => {
+    return (
+      <Breadcrumb {...args} isCollapse={true} itemsBeforeCollapse={2} itemsAfterCollapse={1}>
```
[the remark this answers is not in the document]
Comment (Zyruks) [reply src/components/Breadcrumbs/Breadcrumb.stories.tsx:132]: Done!

File: src/components/Breadcrumbs/Breadcrumb.stories.tsx [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+        <BreadcrumbItem href="javascript:void(0)">Calendar</BreadcrumbItem>
+      </Breadcrumb>
+    );
+  },
+};
+
+export const CollapsedSpread: Story = {
+  render: (args) => {
+    return (
+      <Breadcrumb {...args} isCollapse={true} collapseMode={CollapseMode.spread}>
```
[the remark this answers is not in the document]
Comment (Zyruks) [reply src/components/Breadcrumbs/Breadcrumb.stories.tsx:116]: Done!

File: src/components/Breadcrumbs/Breadcrumb.stories.tsx [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+        </div>
+      </div>
+    );
+  },
+};
+
+export const CollapseDropdown: Story = {
+  render: (args) => {
+    return (
+      <Breadcrumb {...args} isCollapse={true} collapseMode={CollapseMode.dropdown}>
```
[the remark this answers is not in the document]
Comment (Zyruks) [reply src/components/Breadcrumbs/Breadcrumb.stories.tsx:100]: Done!

File: src/components/Breadcrumbs/Breadcrumb.tsx [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+import { cn } from '@common';
+import { CollapseDropdown, CollapsedSpread } from '@components';
+import { ItemIdentifier, processBreadcrumbItem } from './utils/helper';
+
+export enum CollapseMode {
+  dropdown = 'dropdown',
+  spread = 'spread',
+}
+
+export enum BreadcrumbSpacing {
```
[the remark this answers is not in the document]
Comment (Zyruks) [reply src/components/Breadcrumbs/Breadcrumb.tsx:11]: Spacing es el gap entre elementos.

¿Debería tener otro nombre?

image

File: src/components/Breadcrumbs/BreadcrumbCollapsed/CollapsedSpread.tsx [hunk 9]
```diff
@@ -0,0 +1,59 @@
+import React, { ReactNode } from 'react';
```
[the remark this answers is not in the document]
Comment (Zyruks) [reply src/components/Breadcrumbs/BreadcrumbCollapsed/CollapsedSpread.tsx:1]: Done!

File: tailwind.config.js (hunk 2 above)
Comment (Zyruks) [reply tailwind.config.js:9]: De un archivo en el root `tailwind-presets/storybook-preset.js`

image

File: src/components/Breadcrumbs/Breadcrumb.tsx [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+  const childrenAfterCollapse =
+    itemsAfterCollapse >= childrenArray.length
+      ? childrenArray.slice(1)
+      : itemsAfterCollapse <= 0
+      ? childrenArray.slice(-1)
+      : childrenArray.slice(-Math.max(1, itemsAfterCollapse));
+  const collapsedChildren = childrenArray.slice(
+    childrenBeforeCollapse.length,
+    -childrenAfterCollapse.length,
+  );
```
[the remark this answers is not in the document]
Comment (Zyruks) [reply src/components/Breadcrumbs/Breadcrumb.tsx:102]: Done!

File: src/components/Breadcrumbs/BreadcrumbCollapsed/CollapsedSpread.tsx [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+      'cursor-pointer transition-colors',
+    ),
+    collapsedContainer: cn('flex items-center gap-2', '-translate-x-full transition-all', {
+      'absolute left-0 last:left-20 whitespace-nowrap opacity-0 invisible': !isActive,
+      'opacity-100 translate-x-0': isActive,
+    }),
+  };
+  const renderCollapsibleItems = () =>
+    collapsedItems.map((item, index) => (
+      <React.Fragment key={index}>
```
[the remark this answers is not in the document]
Comment (Zyruks) [reply src/components/Breadcrumbs/BreadcrumbCollapsed/CollapsedSpread.tsx:43]: Done!

File: src/components/Breadcrumbs/Breadcrumb.tsx [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+        />
+      );
+    }
+  };
+
+  const renderBreadcrumbs = () => {
+    const effectiveItemsBeforeCollapse = Math.max(1, itemsBeforeCollapse);
+    const effectiveItemsAfterCollapse = Math.max(1, itemsAfterCollapse);
+
+    if (
```
[the remark this answers is not in the document]
Comment (Zyruks) [reply src/components/Breadcrumbs/Breadcrumb.tsx:183]: Done!

File: src/components/Breadcrumbs/Breadcrumb.tsx [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+      'flex items-center',
+      'px-4 py-2 w-fit overflow-scroll',
+      'rounded-xl',
+      BreadcrumbSpacings[spacing],
+      className,
+    ),
+    list: cn('flex items-center', [BreadcrumbSpacings[spacing]]),
+  };
+
+  const startItems = childrenBeforeCollapse.map((item) =>
```
[the remark this answers is not in the document]
Comment (Zyruks) [reply src/components/Breadcrumbs/Breadcrumb.tsx:119]: Done!

File: src/components/Breadcrumbs/Breadcrumb.tsx [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+    processBreadcrumbItem({
+      item,
+      isFirst: item === firstItem,
+      isLast: item === lastItem,
+      separator,
+    }),
+  );
+
+  const collapseItems = collapsedChildren.map((item) =>
```
[the remark this answers is not in the document]
Comment (Zyruks) [reply src/components/Breadcrumbs/Breadcrumb.tsx:128]: Done!

File: src/components/Breadcrumbs/Breadcrumb.tsx [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+    processBreadcrumbItem({
+      collapse: collapseMode === CollapseMode.dropdown,
+      item,
+      separator,
+    }),
+  );
+
+  const endItems = childrenAfterCollapse.map((item) =>
```
[the remark this answers is not in the document]
Comment (Zyruks) [reply src/components/Breadcrumbs/Breadcrumb.tsx:136]: Done!

File: src/components/Breadcrumbs/utils/helper.tsx (hunk 1 above)
Comment (Zyruks) [reply src/components/Breadcrumbs/utils/helper.tsx:11]: Done!

se movio a un customHook `useBreadcrumb`

File: src/hooks/useBreadcrumb/useBreadcrumb.tsx [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+        }),
+      };
+
+      const element = cloneElement(item, { isLast, identifier, ...item.props });
+      const result: Array<ReactNode> = [element];
+
+      switch (true) {
+        case isFirst && identifier === BreadcrumbItemIdentifier.after:
+          result.unshift(
+            <span key="first-separator" className={classes.separator}>
```
Comment (serudda) [inline src/hooks/useBreadcrumb/useBreadcrumb.tsx:61]: No Don Zyruks... hay una confusion aqui... este "custom hook" en realidad te esta devolviendo un componente. Y estoy viendo que te esta devolviendo un "Breadcrumb" (1 solo, multiples son "Breadcrumbs")...

Asi que lo mejor es que se aun component local llamado `<Breadcrumb />`

**Review by serudda — CHANGES_REQUESTED** [review]

File: src/components/Breadcrumbs/Breadcrumb.stories.tsx [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+  BreadcrumbItem,
+  BreadcrumbSpacing,
+  CollapseMode,
+  Icon,
+  IconCatalog,
+} from '@components';
+import { type Meta, type StoryObj } from '@storybook/react';
+
+const meta: Meta<typeof Breadcrumb> = {
+  title: 'UI Components/Breadcrumb',
```
[the remark this answers is not in the document]
Comment (Zyruks) [reply src/components/Breadcrumbs/Breadcrumb.stories.tsx:12]: Done!

File: src/components/Breadcrumbs/Breadcrumb.stories.tsx [hunk 18]
```diff
@@ -0,0 +1,180 @@
...
+  },
+};
+
+export default meta;
+type Story = StoryObj<typeof meta>;
+
+export const Default: Story = {
+  render: (args) => {
+    return (
+      <Breadcrumb {...args}>
```
[the remark this answers is not in the document]
Comment (Zyruks) [reply src/components/Breadcrumbs/Breadcrumb.stories.tsx:32]: Done!

File: src/components/Breadcrumbs/Breadcrumb.tsx [hunk 19]
```diff
@@ -0,0 +1,239 @@
...
+export const Breadcrumb = ({
+  children,
+  className,
+  collapseMode = CollapseMode.spread,
+  isCollapse = false,
+  itemsBeforeCollapse = 1,
+  itemsAfterCollapse = 1,
+  separator = '/',
+  spacing = BreadcrumbSpacing.normal,
+}: BreadcrumbProps) => {
```
[the remark this answers is not in the document]
Comment (Zyruks) [reply src/components/Breadcrumbs/Breadcrumb.tsx:93]: Done!

File: src/components/Breadcrumbs/Breadcrumb.tsx [hunk 20]
```diff
@@ -0,0 +1,239 @@
...
+   * Defines the spacing between breadcrumb items.
+   */
+  spacing?: BreadcrumbSpacing;
+}
+
+/**
+ * Breadcrumbs provide a navigation trail for users to follow back to the starting or entry point of a website or application.
+ * They offer a hierarchical structure of the current page in relation to the website's structure and help users understand their current location.
+ */
+export const Breadcrumb = ({
```
[the remark this answers is not in the document]
Comment (Zyruks) [reply src/components/Breadcrumbs/Breadcrumb.tsx:84]: Done!

File: src/hooks/useBreadcrumb/useBreadcrumb.tsx (hunk 16 above)
Comment (Zyruks) [reply src/hooks/useBreadcrumb/useBreadcrumb.tsx]: Ahora vive dentro de `Breadcrumbs` como function `parseBreadcrumb`

File: src/components/Breadcrumbs/Breadcrumb.tsx (hunk 19 above)
[the remark this answers is not in the document]
Comment (Zyruks) [reply src/components/Breadcrumbs/Breadcrumb.tsx]: Done!

**Review by serudda — APPROVED** [review]

**Pull request merged**